README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->
# Flutter Paypal Pay

## Features
## Upgrade to PayPal v2 API: A Modern Solution

In this sdk, I've transitioned from using the deprecated PayPal v1 API to the latest and most advanced PayPal v2 API. This shift represents a significant improvement in my package, ensuring the utmost security, reliability, and compatibility with current industry standards.

### Key Highlights

- **Upgrade to PayPal v2 API:** I've embraced the latest PayPal v2 API, which offers enhanced functionality, improved security, and a more robust feature set compared to the outdated v1 API.

- **Creating Orders with Ease:** My package now leverages the power of the v2 API to seamlessly create orders, making the process more efficient and user-friendly.

- **Effortless Order Capture:** With the v2 API integration, order capture has become even more straightforward and reliable, ensuring that your transactions are processed smoothly.

- **Stay Current and Compliant:** By using the latest API version, I ensure that my package remains compliant with PayPal's evolving standards, reducing the risk of potential issues or disruptions in your payment processing.

- **Future-Proof Your Application:** With the v2 API, you can future-proof your application, as it aligns with PayPal's ongoing development efforts and ensures that you'll continue to receive support and updates.

Upgrade your package today to take advantage of these significant enhancements and enjoy a more secure and efficient payment processing experience.

#### Recommended usage
If you're selling digital products or services within your app, you can use this SDK to process payments via paypal.

## Installation
```sh
flutter pub add flutter_paypal_pay
```
## How to use the library
```dart

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Center(
        child: ElevatedButton(
          onPressed: () {
            Navigator.of(context).push(
              MaterialPageRoute(
                builder: (BuildContext context) => PaypalPay(

                    /// sandboxMode true for testing purposes
                    sandboxMode: true,

                    /// you will find the client Id after login to paypal developer account .
                    clientId:
                        'AUBmOxdWgvJkGsRf2GZMO96quth3L53nF0B-KBpT5C3JgiJrvifwZuoOMrAkyBWZpw_UaYLVTH162brY',

                    ///you will find the secret key after login to paypal developer account .
                    secretKey:
                        'EBQfKl5oAPlCsGADhNCFe7T8m55PvK1XJO9RgJCjZXpEHfVsLhYhcyjDVAlPe9eF_UW1eJgJN7jb8-Ai',
                    returnURL: 'https://test.com/return',
                    cancelURL: 'https://test.com/cancel',
                    purchaseUnits: const [
                      {
                        'amount': {
                          'value': '100',
                          'currency_code': 'USD',
                        },
                        'shipping': {
                          'address': {
                            // Add your shipping address details here

                            'recipient_name': 'Name',
                            'line1': 'Travis County',
                            'line2': '',
                            'city': 'Austin',
                            'country_code': 'US',
                            'postal_code': '73301',
                            'phone': '+00000000',
                            'state': 'Texas',
                            'admin_area_2':
                                'City Name', // Replace 'City Name' with the actual city or locality name
                            'admin_area_1':
                                'State/Province', // Replace 'State/Province' with the actual state or province name
                          }
                        }
                      }
                    ],
                    note: 'Contact us for any questions on your order.',
                    onSuccess: (Map params) async {
                      //for view the response
                      //response data can be usef for backend validation
                      log('successsss: $params');
                    },
                    onError: (error) {
                      //for view the response
                      log('onEEEEError: $error');
                    },
                    onCancel: (params) {
                      //for view the response
                      log('cancelled: $params');
                    }),
              ),
            );
          },
          child: const Text('Pay with Paypal'),
        ),
      ),
    );
  }
}
```

## Additional information

TODO: Go to https://developer.paypal.com/ and create account the you will get **clientId** and **secretKey** use that to your App. 



CHANGELOG.md
## 0.0.1
* initial release.
## 0.0.2
* minor improvement

## 0.0.3
* init

## 0.0.4

## 0.0.5
* lower http support


lib/src/network_error.dart
import 'package:flutter/material.dart';

class NetworkError extends StatelessWidget {
  final Function loadData;
  final String message;
  final bool isSmall;

  const NetworkError(
      {super.key,
      required this.loadData,
      required this.message,
      this.isSmall = false});

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Image.asset(
            "lib/src/assets/cloud.svg",
            package: "flutter_paypal",
            height: 120,
          ),
          SizedBox(
            height: isSmall ? 20 : 40,
          ),
          Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Text(message,
                  style: const TextStyle(
                      fontSize: 14,
                      color: Color(0xFF272727),
                      fontWeight: FontWeight.w400)),
              const SizedBox(
                width: 5,
              ),
              InkWell(
                onTap: () => loadData(),
                child: const Text("Tap to retry",
                    style: TextStyle(
                        fontSize: 14,
                        color: Colors.black,
                        fontWeight: FontWeight.w800)),
              ),
            ],
          ),
        ],
      ),
    );
  }
}


lib/flutter_paypal_pay.dart
import 'package:flutter/material.dart';
import 'package:flutter_paypal_pay/src/complete_screen.dart';
import 'package:flutter_paypal_pay/src/network_error.dart';
import 'package:flutter_paypal_pay/src/paypal_services.dart';
import 'package:flutter_spinkit/flutter_spinkit.dart';
import 'package:webview_flutter/webview_flutter.dart';
// Import for Android features.
import 'package:webview_flutter_android/webview_flutter_android.dart';
// Import for iOS features.
import 'package:webview_flutter_wkwebview/webview_flutter_wkwebview.dart';

class PaypalPay extends StatefulWidget {
  final Function onSuccess, onCancel, onError;
  final String returnURL, cancelURL, note, clientId, secretKey;
  final List purchaseUnits;
  final bool sandboxMode;
  const PaypalPay({
    Key? key,
    required this.onSuccess,
    required this.onError,
    required this.onCancel,
    required this.returnURL,
    required this.cancelURL,
    required this.purchaseUnits,
    required this.clientId,
    required this.secretKey,
    this.sandboxMode = false,
    this.note = '',
  }) : super(key: key);

  @override
  State<StatefulWidget> createState() {
    return PaypalPayState();
  }
}

class PaypalPayState extends State<PaypalPay> {
  late final WebViewController _controller;
  String checkoutUrl = '';
  String navUrl = '';
  String executeUrl = '';
  String accessToken = '';
  bool loading = true;
  bool pageLoading = true;
  bool loadingError = false;
  late PaypalServices services;
  int pressed = 0;

  Map getOrderParams() {
    Map<String, dynamic> temp = {
      'intent': 'CAPTURE',
      'purchase_units': widget.purchaseUnits,
      'payment_source': {
        'paypal': {
          'experience_context': {
            'payment_method_preference': 'IMMEDIATE_PAYMENT_REQUIRED',
            'brand_name': 'EXAMPLE INC',
            'locale': 'en-US',
            'landing_page': 'LOGIN',
            'shipping_preference': 'SET_PROVIDED_ADDRESS',
            'user_action': 'PAY_NOW',
            'return_url': widget.returnURL,
            'cancel_url': widget.cancelURL,
          }
        }
      }
    };
    return temp;
  }

  loadPayment() async {
    setState(() {
      loading = true;
    });
    try {
      Map getToken = await services.getAccessToken();
      if (getToken['token'] != null) {
        accessToken = getToken['token'];
        final transactions = getOrderParams();
        final res =
            await services.createPaypalPayment(transactions, accessToken);
        if (res['approvalUrl'] != null) {
          setState(() {
            checkoutUrl = res['approvalUrl'].toString();
            navUrl = res['approvalUrl'].toString();
            executeUrl = res['executeUrl'].toString();
            loading = false;
            pageLoading = false;
            loadingError = false;
          });
          _controller.loadRequest(Uri.parse(checkoutUrl));
        } else {
          widget.onError(res);
          setState(() {
            loading = false;
            pageLoading = false;
            loadingError = true;
          });
        }
      } else {
        widget.onError("${getToken['message']}");

        setState(() {
          loading = false;
          pageLoading = false;
          loadingError = true;
        });
      }
    } catch (e) {
      widget.onError(e);
      setState(() {
        loading = false;
        pageLoading = false;
        loadingError = true;
      });
    }
  }

  @override
  void initState() {
    super.initState();
    services = PaypalServices(
      sandboxMode: widget.sandboxMode,
      clientId: widget.clientId,
      secretKey: widget.secretKey,
    );
    setState(() {
      navUrl = widget.sandboxMode
          ? 'https://api-m.sandbox.paypal.com'
          : 'https://www.api.paypal.com';
    });
    // Enable hybrid composition.
    loadPayment();

    // #docregion platform_features
    late final PlatformWebViewControllerCreationParams params;
    if (WebViewPlatform.instance is WebKitWebViewPlatform) {
      params = WebKitWebViewControllerCreationParams(
        allowsInlineMediaPlayback: true,
        mediaTypesRequiringUserAction: const <PlaybackMediaTypes>{},
      );
    } else {
      params = const PlatformWebViewControllerCreationParams();
    }

    final WebViewController controller =
        WebViewController.fromPlatformCreationParams(params);
    // #enddocregion platform_features

    controller
      ..setJavaScriptMode(JavaScriptMode.unrestricted)
      ..setBackgroundColor(const Color(0x00000000))
      ..setNavigationDelegate(
        NavigationDelegate(
          onProgress: (int progress) {
            debugPrint('WebView is loading (progress : $progress%)');
          },
          onPageStarted: (String url) {
            setState(() {
              pageLoading = true;
              loadingError = false;
            });
            debugPrint('Page started loading: $url');
          },
          onPageFinished: (String url) {
            setState(() {
              navUrl = url;
              pageLoading = false;
            });
          },
          onWebResourceError: (WebResourceError error) {
            debugPrint('''
              Page resource error:
              code: ${error.errorCode}
              description: ${error.description}
              errorType: ${error.errorType}
              isForMainFrame: ${error.isForMainFrame}
          ''');
          },
          onNavigationRequest: (NavigationRequest request) async {
            if (request.url.startsWith('https://www.youtube.com/')) {
              debugPrint('blocking navigation to ${request.url}');
              return NavigationDecision.prevent;
            }
            if (request.url.contains(widget.returnURL)) {
              Navigator.pushReplacement(
                context,
                MaterialPageRoute(
                    builder: (context) => CompletePayment(
                        url: request.url,
                        services: services,
                        executeUrl: executeUrl,
                        accessToken: accessToken,
                        onSuccess: widget.onSuccess,
                        onCancel: widget.onCancel,
                        onError: widget.onError)),
              );
            }
            if (request.url.contains(widget.cancelURL)) {
              final uri = Uri.parse(request.url);
              await widget.onCancel(uri.queryParameters);
              // ignore: use_build_context_synchronously
              Navigator.of(context).pop();
            }
            debugPrint('allowing navigation to ${request.url}');
            return NavigationDecision.navigate;
          },
          onUrlChange: (UrlChange change) {
            debugPrint('url change to ${change.url}');
          },
        ),
      )
      ..addJavaScriptChannel(
        'Toaster',
        onMessageReceived: (JavaScriptMessage message) {
          ScaffoldMessenger.of(context).showSnackBar(
            SnackBar(content: Text(message.message)),
          );
        },
      );

    // #docregion platform_features
    if (controller.platform is AndroidWebViewController) {
      AndroidWebViewController.enableDebugging(true);
      (controller.platform as AndroidWebViewController)
          .setMediaPlaybackRequiresUserGesture(false);
    }
    // #enddocregion platform_features

    _controller = controller;
  }

  @override
  Widget build(BuildContext context) {
    return WillPopScope(
      onWillPop: () async {
        if (pressed < 2) {
          setState(() {
            pressed++;
          });
          final snackBar = SnackBar(
              content: Text(
                  'Press back ${3 - pressed} more times to cancel transaction'));
          ScaffoldMessenger.of(context).showSnackBar(snackBar);
          return false;
        } else {
          return true;
        }
      },
      child: Scaffold(
          appBar: AppBar(
            backgroundColor: Colors.blueGrey,
            leading: GestureDetector(
              child: const Icon(
                Icons.arrow_back_ios_new,
                color: Colors.white,
              ),
              onTap: () => Navigator.pop(context),
            ),
            title: Row(
              children: [
                Expanded(
                    child: Container(
                  padding:
                      const EdgeInsets.symmetric(horizontal: 16, vertical: 8),
                  decoration: BoxDecoration(
                      color: Colors.black.withOpacity(0.3),
                      borderRadius: BorderRadius.circular(20)),
                  child: Row(
                    children: [
                      Icon(
                        Icons.lock_outline,
                        color: Uri.parse(navUrl).hasScheme
                            ? Colors.green
                            : Colors.blue,
                        size: 18,
                      ),
                      const SizedBox(width: 5),
                      Expanded(
                        child: Text(
                          navUrl,
                          overflow: TextOverflow.ellipsis,
                          style: const TextStyle(fontSize: 14),
                        ),
                      ),
                      SizedBox(width: pageLoading ? 5 : 0),
                      pageLoading
                          ? const SpinKitFadingCube(
                              color: Color(0xFFEB920D),
                              size: 10.0,
                            )
                          : const SizedBox()
                    ],
                  ),
                ))
              ],
            ),
            elevation: 0,
          ),
          body: SizedBox(
            height: MediaQuery.of(context).size.height,
            width: MediaQuery.of(context).size.width,
            child: loading
                ? const Column(
                    children: [
                      Expanded(
                        child: Center(
                          child: SpinKitRotatingCircle(
                            color: Color(0xFFEB920D),
                            size: 30.0,
                          ),
                        ),
                      ),
                    ],
                  )
                : loadingError
                    ? Column(
                        children: [
                          Expanded(
                            child: Center(
                              child: NetworkError(
                                  loadData: loadPayment,
                                  message: 'Something went wrong,'),
                            ),
                          ),
                        ],
                      )
                    : Column(
                        children: [
                          Expanded(
                            child: WebViewWidget(controller: _controller),
                          ),
                        ],
                      ),
          )),
    );
  }
}


lib/src/complete_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_paypal_pay/src/network_error.dart';
import 'package:flutter_paypal_pay/src/paypal_services.dart';
import 'package:flutter_spinkit/flutter_spinkit.dart';
// ignore_for_file: use_build_context_synchronously

class CompletePayment extends StatefulWidget {
  final Function onSuccess, onCancel, onError;
  final PaypalServices services;
  final String url, executeUrl, accessToken;
  const CompletePayment({
    Key? key,
    required this.onSuccess,
    required this.onError,
    required this.onCancel,
    required this.services,
    required this.url,
    required this.executeUrl,
    required this.accessToken,
  }) : super(key: key);

  @override
  // ignore: library_private_types_in_public_api
  _CompletePaymentState createState() => _CompletePaymentState();
}

class _CompletePaymentState extends State<CompletePayment> {
  bool loading = true;
  bool loadingError = false;

  complete() async {
    final uri = Uri.parse(widget.url);
    final payerID = uri.queryParameters['PayerID'];
    if (payerID != null) {
      Map params = {
        'payerID': payerID,
        'token': uri.queryParameters['token'],
      };
      setState(() {
        loading = true;
        loadingError = false;
      });

      Map resp = await widget.services.executePayment(widget.accessToken);
      if (resp['error'] == false) {
        params['status'] = 'success';
        params['data'] = resp['data'];
        await widget.onSuccess(params);
        setState(() {
          loading = false;
          loadingError = false;
        });
        Navigator.pop(context, params);
      } else {
        if (resp['exception'] != null && resp['exception'] == true) {
          widget.onError({'message': resp['message']});
          setState(() {
            loading = false;
            loadingError = true;
          });
        } else {
          await widget.onError(resp['data']);
          Navigator.of(context).pop();
        }
      }
      //return NavigationDecision.prevent;
    } else {
      Navigator.of(context).pop();
    }
  }

  @override
  void initState() {
    super.initState();
    complete();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        child: loading
            ? const Column(
                children: [
                  Expanded(
                    child: Center(
                      child: SpinKitFadingCube(
                        color: Color(0xFFEB920D),
                        size: 30.0,
                      ),
                    ),
                  ),
                ],
              )
            : loadingError
                ? Column(
                    children: [
                      Expanded(
                        child: Center(
                          child: NetworkError(
                              loadData: complete,
                              message: 'Something went wrong,'),
                        ),
                      ),
                    ],
                  )
                : const Center(
                    child: Text('Payment Completed'),
                  ),
      ),
    );
  }
}


lib/src/paypal_services.dart
import 'dart:async';
import 'dart:convert' as convert;

import 'package:http/http.dart' as http;
import 'package:http_auth/http_auth.dart';

class PaypalServices {
  String id = '';
  final String clientId, secretKey;
  final bool sandboxMode;
  PaypalServices({
    required this.clientId,
    required this.secretKey,
    required this.sandboxMode,
  });

  getAccessToken() async {
    String domain = sandboxMode
        ? 'https://api.sandbox.paypal.com'
        : 'https://api.paypal.com';
    try {
      var client = BasicAuthClient(clientId, secretKey);
      var response = await client.post(
          Uri.parse('$domain/v1/oauth2/token?grant_type=client_credentials'));
      if (response.statusCode == 200) {
        final body = convert.jsonDecode(response.body);
        return {
          'error': false,
          'message': 'Success',
          'token': body['access_token']
        };
      } else {
        return {
          'error': true,
          'message': 'Your PayPal credentials seems incorrect'
        };
      }
    } catch (e) {
      return {
        'error': true,
        'message': 'Unable to proceed, check your internet connection.'
      };
    }
  }

  Future<Map> createPaypalPayment(transactions, accessToken) async {
    String domain = sandboxMode
        ? 'https://api-m.sandbox.paypal.com'
        : 'https://api.paypal.com';
    try {
      var response = await http.post(Uri.parse('$domain/v2/checkout/orders'),
          body: convert.jsonEncode(transactions),
          headers: {
            'content-type': 'application/json',
            'Authorization': 'Bearer $accessToken'
          });

      final body = convert.jsonDecode(response.body);
      if (response.statusCode == 200) {
        id = body['id'];
        if (body['links'] != null && body['links'].length > 0) {
          List links = body['links'];

          //String executeUrl = '';
          String approvalUrl = '';
          final item = links.firstWhere((o) => o['rel'] == 'payer-action',
              orElse: () => null);
          if (item != null) {
            approvalUrl = item['href'];
          }
          /* final item1 =
              links.firstWhere((o) => o['rel'] == 'self', orElse: () => null);
          if (item1 != null) {
            executeUrl = item1['href'];
          }
           */
          return {'approvalUrl': approvalUrl};
        }
        return {};
      } else {
        return body;
      }
    } catch (e) {
      rethrow;
    }
  }

  Future<Map> executePayment(accessToken) async {
    String domain = sandboxMode
        ? 'https://api-m.sandbox.paypal.com'
        : 'https://api.paypal.com';
    try {
      var response = await http
          .post(Uri.parse('$domain/v2/checkout/orders/$id/capture'), headers: {
        'content-type': 'application/json',
        'Authorization': 'Bearer $accessToken'
      });

      final body = convert.jsonDecode(response.body);
      if (response.statusCode == 201) {
        return {'error': false, 'message': 'Success', 'data': body};
      } else {
        return {
          'error': true,
          'message': 'Payment inconclusive.',
          'data': body
        };
      }
    } catch (e) {
      return {'error': true, 'message': e, 'exception': true, 'data': null};
    }
  }
}


example/lib/main.dart
import 'dart:developer';

import 'package:flutter/material.dart';
import 'package:flutter_paypal_pay/flutter_paypal_pay.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Paypal Pay Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Center(
        child: ElevatedButton(
          onPressed: () {
            Navigator.of(context).push(
              MaterialPageRoute(
                builder: (BuildContext context) => PaypalPay(

                    /// sandboxMode true for testing purposes
                    sandboxMode: true,

                    /// you will find the client Id after login to paypal developer account .
                    clientId:
                        'AUBmOxdWgvJkGsRf2GZMO96quth3L53nF0B-KBpT5C3JgiJrvifwZuoOMrAkyBWZpw_UaYLVTH162brY',

                    ///you will find the secret key after login to paypal developer account .
                    secretKey:
                        'EBQfKl5oAPlCsGADhNCFe7T8m55PvK1XJO9RgJCjZXpEHfVsLhYhcyjDVAlPe9eF_UW1eJgJN7jb8-Ai',
                    returnURL: 'https://test.com/return',
                    cancelURL: 'https://test.com/cancel',
                    purchaseUnits: const [
                      {
                        'amount': {
                          'value': '100',
                          'currency_code': 'USD',
                        },
                        'shipping': {
                          'address': {
                            // Add your shipping address details here

                            'recipient_name': 'Name',
                            'line1': 'Travis County',
                            'line2': '',
                            'city': 'Austin',
                            'country_code': 'US',
                            'postal_code': '73301',
                            'phone': '+00000000',
                            'state': 'Texas',
                            'admin_area_2':
                                'City Name', // Replace 'City Name' with the actual city or locality name
                            'admin_area_1':
                                'State/Province', // Replace 'State/Province' with the actual state or province name
                          }
                        }
                      }
                    ],
                    note: 'Contact us for any questions on your order.',
                    onSuccess: (Map params) async {
                      //for view the response
                      //response data can be usef for backend validation
                      log('successsss: $params');
                    },
                    onError: (error) {
                      //for view the response
                      log('onEEEEError: $error');
                    },
                    onCancel: (params) {
                      //for view the response
                      log('cancelled: $params');
                    }),
              ),
            );
          },
          child: const Text('Pay with Paypal'),
        ),
      ),
    );
  }
}
